pallets/pallet-bonded-coins/src/curves.rs

```diff
@@ -0,0 +1,260 @@
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_arithmetic::ArithmeticError;
+use sp_std::ops::{AddAssign, BitOrAssign, ShlAssign};
+use substrate_fixed::{
+	traits::{Fixed, FixedSigned, ToFixed},
+	transcendental::{exp, ln, sqrt},
+};
+
+use crate::{PassiveSupply, Precision};
+
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub enum Curve<Parameter> {
+	Polynomial(PolynomialParameters<Parameter>),
+	SquareRoot(SquareRootParameters<Parameter>),
+	LMSR(LMSRFunctionParameters<Parameter>),
+}
+
+pub enum Operation<PassiveSupply> {
+	Mint(PassiveSupply),
+	Burn(PassiveSupply),
+}
+
+impl<Balance> Operation<PassiveSupply<Balance>> {
+	pub fn inner_value(&self) -> &PassiveSupply<Balance> {
+		match self {
+			Operation::Mint(x) => x,
+			Operation::Burn(x) => x,
+		}
+	}
+}
+
+impl<Parameter> Curve<Parameter>
+where
+	Parameter: FixedSigned + PartialOrd<Precision> + From<Precision> + ToFixed,
+	<Parameter as Fixed>::Bits: Copy + ToFixed + AddAssign + BitOrAssign + ShlAssign,
+{
+	fn calculate_accumulated_passive_issuance<Balance: Fixed>(passive_issuance: &[Balance]) -> Balance {
+		passive_issuance
+			.iter()
+			.fold(Balance::from_num(0), |sum, x| sum.saturating_add(*x))
+	}
+
+	fn calculate_integral_bounds(
+		op: Operation<PassiveSupply<Parameter>>,
+		active_issuance_pre: Parameter,
+		active_issuance_post: Parameter,
+	) -> (Parameter, Parameter) {
+		match op {
+			Operation::Burn(passive) => {
+				let accumulated_passive_issuance = Self::calculate_accumulated_passive_issuance(&passive);
+				(
+					active_issuance_post.saturating_add(accumulated_passive_issuance),
+					active_issuance_pre.saturating_add(accumulated_passive_issuance),
+				)
+			}
+			Operation::Mint(passive) => {
+				let accumulated_passive_issuance = Self::calculate_accumulated_passive_issuance(&passive);
+				(
+					active_issuance_pre.saturating_add(accumulated_passive_issuance),
+					active_issuance_post.saturating_add(accumulated_passive_issuance),
+				)
+			}
+		}
+	}
+	pub fn calculate_cost(
+		&self,
+		active_issuance_pre: Parameter,
+		active_issuance_post: Parameter,
+		op: Operation<PassiveSupply<Parameter>>,
```

Comment:
My preference would be to have
```suggestion
		&self,
		active_issuance_lower: Parameter,
		active_issuance_higher: Parameter,
		passive_issuance: PassiveSupply<Parameter>,
```
and to let the pallet implementation deal with implementing how to compute active issuances.

Reply:
A secondary question would be if we even need `PassiveSupply<Parameter>` to be a `Vec` or if it should already be summed up before being passed into this function.

Reply:
I think we need it as a vector for the LMSR function. In LMSR, e^passive_supply is not the same as e^supply1 + e^supply2.

Reply:
Please check out this commit: https://github.com/KILTprotocol/kilt-node/pull/770/commits/4431d393209968523b17778570c53fe0a93f4992 

I applied the changes you suggested. I found it much better now!